livre/lib/cache/HistogramObject.cpp

```diff
@@ -0,0 +1,240 @@
+                                (( data / range ) + 0.5f ) * Histogram::binCount;
+                        bins[ binIndex ] += scaleFactor;
+                    }
+    }
+
+    bool load()
+    {
+        const VolumeInformation& volumeInfo = _dataCache.getDataSource().getVolumeInfo();
+        const size_t compCount = volumeInfo.compCount;
+
+        if( compCount > 1 )
+            LBTHROW( std::runtime_error( "Multiple channels are not supported "));
+
+        ConstTextureDataObjectPtr textureData =
+                std::static_pointer_cast< const TextureDataObject >(
+                    _dataCache.get( _histogramObject.getId( )));
+
+        if( !textureData->isLoaded( ))
+            return false;
+
+        const void* rawData = textureData->getDataPtr();
+        const LODNode& lodNode = _dataCache.getDataSource().getNode(
+                    NodeId( _histogramObject.getId( )));
+        const Vector3ui& voxelBox = lodNode.getVoxelBox().getSize();
+        const Vector3ui& padding = volumeInfo.overlap;
+
+        const uint64_t scaleFactor1d =
+                1 << ( volumeInfo.rootNode.getDepth() - lodNode.getRefLevel() - 1 );
+
+        const uint64_t scaleFactor = scaleFactor1d * scaleFactor1d * scaleFactor1d;
+
+        const DataType dataType = volumeInfo.dataType;
+        Histogram histogram;
+        switch( dataType )
+        {
+           case DT_UINT8:
+                binData( static_cast< const uint8_t* >( rawData ), histogram,
+                         voxelBox, padding, compCount, scaleFactor  );
+                break;
+           case DT_UINT16:
+                binData< uint16_t >( static_cast< const uint16_t* >( rawData ), histogram,
+                                     voxelBox, padding, compCount, scaleFactor  );
+                break;
+           case DT_UINT32:
+                binData< uint32_t >( static_cast< const uint32_t* >( rawData ), histogram,
+                                     voxelBox, padding, compCount, scaleFactor  );
+                break;
+           case DT_INT8:
+                binData( static_cast< const int8_t* >( rawData ), histogram,
+                         voxelBox, padding, compCount, scaleFactor  );
+                break;
+           case DT_INT16:
+                binData< int16_t >( static_cast< const int16_t* >( rawData ), histogram,
+                                    voxelBox, padding, compCount, scaleFactor  );
+                break;
+           case DT_INT32:
+                binData< int32_t >( static_cast< const int32_t* >( rawData ), histogram,
+                                    voxelBox, padding, compCount, scaleFactor  );
+                break;
+           case DT_FLOAT32:
+           case DT_FLOAT64:
+           case DT_UNDEFINED:
+           default:
+           {
+                LBTHROW( std::runtime_error( "Unimplemented data type." ));
+           }
+        }
+         _histogram.reset( new Histogram( histogram ));
```

Comment:
I mean storing _histogram by value, or do you transfer the unique_ptr somewhere else?